### PR TITLE
Insight folders

### PR DESCRIPTION
**Pull request by dominikl**
PR for adding ROI folders support to Insight

**Test:** Check that the measurement tool still works with "plain" ROIs as well as with ROIs in folders.

### CODE REVIEW

File: components/blitz/src/ome/services/blitz/impl/RoiI.java [hunk 1]
```diff
@@ -1,7 +1,7 @@
 /*
  *   $Id$
  *
- *   Copyright 2009 Glencoe Software, Inc. All rights reserved.
+ *   Copyright 2009-2016 Glencoe Software, Inc. All rights reserved.
```
Comment (mtbc) [inline components/blitz/src/ome/services/blitz/impl/RoiI.java]: best not to touch Glencoe headers at present

File: components/blitz/src/omero/gateway/model/DataObject.java [hunk 2]
```diff
@@ -511,6 +510,17 @@ public Project asProject() {
     }
 
     /**
+     * Returns the hosted IObject as a Folder. Not null; may through class-cast
```
Comment (mtbc) [inline components/blitz/src/omero/gateway/model/DataObject.java]: throw

File: components/blitz/src/ome/services/blitz/impl/RoiI.java (hunk 1 above)
Comment (dominikl) [inline components/blitz/src/ome/services/blitz/impl/RoiI.java]: No, that was me; just became a habit to adjust copyright date. Will revert that.

File: components/insight/SRC/org/openmicroscopy/shoola/agents/measurement/view/ROITable.java [hunk 3]
```diff
@@ -491,11 +539,58 @@ void removeROI(ROI roi)
...
+     *            The collection to search through
+     * @param folder
+     *            The folder to look for
+     * @return The ROINode representing the folder or <code>null</code> if it
+     *         can't be found
+     */
+    private ROINode findFolderNode(Collection<ROINode> nodes, FolderData folder) {
+        for (ROINode n : nodes) {
+            Object obj = n.getUserObject();
+            if (obj != null && obj instanceof FolderData
```
Comment (mtbc) [inline components/insight/SRC/org/openmicroscopy/shoola/agents/measurement/view/ROITable.java]: if it's null it can't be instanceof FolderData

**Comment by sbesson** [comment]
See http://regions-ci.docker.openmicroscopy.org:8080/job/OMERO-start/. The integration tests now systematically fail `OmeroPy.test.integration.test_rois.TestRois.test8990`. Not sure if this related to this PR /cc @mtbc

**Comment by mtbc** [comment]
It might be: on my local folders branch I get,
```
============================= test session starts ==============================
platform linux2 -- Python 2.7.9 -- py-1.4.25 -- pytest-2.6.3 -- /usr/bin/python
collecting ... collected 4 items
test/integration/test_rois.py::TestRois::teststats1 PASSED
test/integration/test_rois.py::TestRois::test3703 PASSED
test/integration/test_rois.py::TestRois::testGetROICount PASSED
test/integration/test_rois.py::TestRois::test8990 PASSED
 generated xml file: /home/mtbc/src/openmicroscopy/components/tools/OmeroPy/target/reports/junit-results.xml 
=========================== 4 passed in 7.67 seconds ===========================
```
so my PRs are *probably* okay.

**Comment by mtbc** [comment]
Tested also with exactly #4441 and #4445 merged onto regions, still 4 passes.

**Comment by dominikl** [comment]
Interesting... yes, looks like it's my PR. I'll investigate and add a respective java integration test.

**Comment by sbesson** [comment]
screen shot 2016-02-02 at 13 43 06

File: components/insight/SRC/org/openmicroscopy/shoola/agents/measurement/util/roitable/ROITableModel.java [hunk 4]
```diff
@@ -69,7 +69,7 @@
 
 	/** Visible Column no for the wizard. */
 	public static final int				VISIBLE_COLUMN = 5;
-
+
```
Comment (mtbc) [inline components/insight/SRC/org/openmicroscopy/shoola/agents/measurement/util/roitable/ROITableModel.java]: probably accidental

**Comment by gusferguson** [comment]
@dominikl cc @pwalczysko 

Tested with regions-omero.docker.openmicroscopy.org sample_user

Behaves as expected.
Screenshot done.
Good to merge.

**Comment by pwalczysko** [comment]
As discussed this morning, additional testing images:
  - big image (suggest ``test_images_good/svs/alexandra/UMD001_ORO.svs``)
  - multi-t & mulit-z image (suggest ``test_images_good/zeiss-lsm/2chZT.lsm``)
  - plate (suggest ``test_images_good/bd-pathway/``

**Comment by pwalczysko** [comment]
This was not the topic of this PR, but still, need to take care of that in next steps : When selecting a ROI which is included in 2 folders, this is being selected twice in the ROI Manager. As a result, such ROI cannot be propagated, because the UI will answer ``can propagate just one ROI at a time``.

**Comment by pwalczysko** [comment]
Considering the https://github.com/openmicroscopy/openmicroscopy/pull/4450#issuecomment-179268961 to be for the next PRs (stages), this is ready to merge, as the normal manipulation with ROIs on the images went fine.

**Comment by gusferguson** [comment]
@pwalczysko - will add this to the issues so it doesn't get missed.

**Comment by pwalczysko** [comment]
Another comment for follow-up work: Visibility of ROIs which are contained in more than one folder:

   - have 2 ROIs in a folder (r0 and r1 in f1)
   - have 2 ROIs in another folder (r0 and r2 in f2)
   - this means that ROI r1 is  contained in both folders
   - ROIs r1 and r2 are specific (for folders f1 and f2 respectively)
   - start with all ROIs and folders visible (f1, f2 as well as r0, r1 and r2 are all ticked visible)
   - now deselect folder f1 -> observe that the r0 region became invisible, although it is still contained in the other folder f2, which is still visible
   - my suggestion would be: if the ROI is contained in some other, visible folder, then leave it visible, even as the folder f1 (which contains the ROI as well) is invisible

**Comment by sbesson** [comment]
Any veto to getting this merged and handling the visibility, the next step (nested folders) and Mark's comment in follow-up PRs?

**Comment by mtbc** [comment]
Can ignore my comment if in followup PR as was mostly about keeping diff clean.

**Comment by dominikl** [comment]
That's because in the current implementation the "Visible" status of the ROI folder is determined by the Visible status of its ROIs, and the change of the Visible status of the ROI folder is propagated to all its ROIs. But as discussed with @pwalczysko , it has to be the other way round. The Visible status of the ROI has to take the Visible status of all it's parent Folders into account.

Yes, I'd suggest to tackle this in the next follow up PR.

**Comment by gusferguson** [comment]
Above comments added to https://github.com/openmicroscopy/design/issues/11

**Comment by jburel** [comment]
@sbesson: FYI some typo fixed in this PR. It is already merged but hopefully no conflict when porting to develop

**Comment by jburel** [comment]
@dominikl: a null check should have been added when setting name/description in ```FolderData```

**Comment by jburel** [comment]
@dominikl: In ROITest better to use ```assertEquals``` than ```assertTrue```. easier to identify error

**Comment by jburel** [comment]
@dominikl: ``` if (folders != null) newROI.getFolders().addAll(folders); ```
also check if list is empty

**Comment by sbesson** [comment]
@jburel: true, in `DataObject.java` mostly. As we are weekly merging `develop` into this branch, we should be able to minimize potential conflicts.

**Pull request merged**